blaze/server/spider.py

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python
+
+from __future__ import absolute_import
+
+import os
+
+from odo import resource
+from odo.utils import ignoring
+
+from .server import Server, DEFAULT_PORT
+
+
+__all__ = 'spider',
+
+
+def _spider(resource_path, ignore, followlinks, hidden):
+    resources = {}
+    for filename in (os.path.join(resource_path, x)
+                     for x in os.listdir(resource_path)):
+        basename = os.path.basename(filename)
+        if (basename.startswith(os.curdir) and not hidden or
+                os.path.islink(filename) and not followlinks):
+            continue
+        if os.path.isdir(filename):
+            new_resources = _spider(filename, ignore=ignore,
+                                    followlinks=followlinks, hidden=hidden)
+            if new_resources:
+                resources[basename] = new_resources
+        else:
+            with ignoring(*ignore):
+                resources[basename] = resource(filename)
+    return resources
+
+
+def spider(resource_path, ignore=(ValueError, NotImplementedError),
+           followlinks=True, hidden=False):
+    return {
+        os.path.basename(resource_path): _spider(resource_path, ignore=ignore,
+                                                 followlinks=followlinks,
+                                                 hidden=hidden)
+    }
+
+
+def from_yaml(filename, ignore=(ValueError, NotImplementedError),
+              followlinks=True, hidden=False):
+    import yaml
+    with open(filename, 'rt') as f:
+        spec = yaml.load(f.read())
+    resources = {}
+    for name, info in spec.items():
+        if 'uri' not in info:
+            raise ValueError('uri not found for data source %r' % name)
+        uri = info['uri']
+        if os.path.isdir(uri):
+            resources = spider(os.path.expanduser(uri),
+                               ignore=ignore,
+                               followlinks=followlinks,
+                               hidden=hidden)
+        else:
+            resources[name] = resource(uri, dshape=info.get('dshape'))
+    return resources
+
+
+def _parse_args():
```

Comment:
> Should this accept a `host` argument to pass to the server?

It does, via the `--host` argument on the command line

> Also, on the line below, is there a reason for the call-time import?

`import odo` is somewhat slow because of the many `dispatch` decorated functions. I was trying to avoid importing odo until needed so that `--help` would be relatively quick, but since it causes other more insidious problems I'll move them to the top level of the module.